gossip/src/crds_value.rs

```diff
@@ -38,6 +38,8 @@ pub const MAX_VOTES: VoteIndex = 32;
 
 pub type EpochSlotsIndex = u8;
 pub const MAX_EPOCH_SLOTS: EpochSlotsIndex = 255;
+// We now keep 81000 slots, 81000/MAX_SLOTS_PER_ENTRY = 5.
```

Comment:
Where is this `81000` coming from?
These CRDS values aren't really cheap. Do we really need this many?

Reply:
The number comes from https://github.com/solana-foundation/solana-improvement-documents/pull/46.

During an outage we don't know each other's status, so there is no way to know how lagged others are. So we send out 9 hours of slots, hoping that we can find out about an outage in 9 hours.

The values are big, but we thought it was okay because these messages are only transmitted during an outage. Later after I add the command line --wen_restart flag I can come back and filter out these CRDS values during non-restart mode if that makes you feel safer.

Reply:
`81_000` slots is ~5MB of data _per node_ in the cluster!!!

Also, I am not sure if a node is `81_000` slots behind other nodes, it can repair all those slots, so what is the point?
I think we really need to revise this number lower.

Reply:
As I understand it, 9 hours is chosen because in the past our outage resolution normally takes 4~7 hours just for the validator restart part. And we add a few hours to be on the cautious side. Not every validator operator has a pager, but most of them will tend to things within 9 hours.

Sending 81k slots on the Gossip messages doesn't mean you need to repair all those slots. Most of them are probably older than your local root so you will instantly dump it on the floor.

The tricky part here is you don't know what the status of others are so you really want to send more than they probably need to make the restart successful.

Reply:
> Sending 81k slots on the Gossip messages doesn't mean you need to repair all those slots. Most of them are probably older than your local root so you will instantly dump it on the floor.

There are 2 different scenarios:
* If a node has some of the intermediate slots, then it can already repair missing slots and complete the chain. No need to send 81k slots to it.
* If a node does not have any of the intermediate slots, do we really expect it to be able to repair 81k slots?

Even before all these, why do we need to send the parent slots to begin with?
We are already able to repair a fork if have some missing parent slots in the fork by sending repair orphan request:
https://github.com/solana-labs/solana/blob/bc2b37276/core/src/repair/serve_repair.rs#L240-L243
Why do we need to duplicate that protocol here? The node can already recover the whole fork by repairing off the last slot.

Reply:
This is outage handling code, so the main theme here is: do the simplest thing possible and reach consensus asap. We've had outage in the past where repair service is moving really slowly due to various reasons. If the repair service was able to repair missing blocks correctly, we might not need this code at all.

So the main protocol here is basically try to spread local voting history metadata asap, then repair only the essential blocks, then try to reach consensus on where to restart from.

Reply:
If you already have some data shreds from the respective slots, then you also know the respective `parent` slot from the data shreds and won't need a long 81k slots fork chain.
If you don't have any shreds from the slot, then you need to repair ~800 data shreds for the slot. Sending one additional repair orphan request:
https://github.com/solana-labs/solana/blob/bc2b37276/core/src/repair/serve_repair.rs#L240-L243
to obtain the parent-chain won't really change anything in practice.

i.e. you are reducing 801 repair requests to 800 at the cost of many more gossip messages. This is clearly negative trade-off.

Reply:
This step is more of doing a quick bean counting to quickly figure out which slots are ones you must repair, which are the ones you can ignore, so the world can quickly converge to a sane state so people can start the next consensus step to say which one is the restart slot the cluster should use.

Think about an outage where tons of small forks are created but none of them can make progress. Sure, I know slot X+1 to slot X+81000 are all child of slot Y, I probably already have some data shreds present to verify the Y -> X+N parent-child relationship, but repairing all of them might take hours.

While if you do a quick bean counting at the very beginning, you could realize no children of slot Y ever got pass 42% vote stake, so you can skip repairing all these children and claim Y is the restart slot everyone should use. Then we can converge on a consensus result much faster.

81000 is a safety measure to make sure we can quickly do this bean counting in all sorts of outage scenarios we never imagined. In reality, we probably don't need 99% of the 81000 slots, but it's hard to know how many you should send when it's hard to talk to your peers, also our local optimistically confirmed slot might be wrong because of software bugs, so this last voted fork is something very basic we can rely on.

Yes I realize this Gossip message might be big, but it's comparable to the size of EpochSlots so it's definitely doable, also that's why I'll put in two checks (shred_version check and wen_restart mode check) to make sure they don't pollute CRDS table of a validator not in wen_restart.

Reply:
I think this design unnecessarily makes things complicated, in addition to duplicating already existing logic in repair service; and I am also not convinced this achieves a significant performance improvement.
Do you want to do some testing first to see how long it takes to recover the parent chain of a fork just using the repair orphan request?
https://github.com/solana-labs/solana/blob/bc2b37276/core/src/repair/serve_repair.rs#L240-L243
note that this request only returns **one** data shred for each slot, which is enough to obtain the parent chain.

Reply:
Of course we can optimize the design, but it's probably better to discuss over the SIMD PR so more people can join the discussion.
https://github.com/solana-foundation/solana-improvement-documents/pull/46

As I said, I do get your point that using this purely for repairing is probably not too much more efficient compared to the current protocol, but we do use this for consensus vote as well, which is something the current repair protocol can't do.

Reply:
By the way, I think 81k slots is less than 5MB, because it's a bitmap. I assume the current MAX_SLOTS_PER_ENTRY is chosen so each EpochSlots piece fits within a 1460 byte UDP packet, and MAX_RESTART_LAST_VOTED_FORK_SLOTS is 5, which is ~7KB per validator?  Let's say it's a 64K UDP packet, that's still smaller than 5MB per validator.

How much memory does the current EpochSlots messages occupy? Considering MAX_EPOCH_SLOTS is 255, this should be much smaller than that.

Is my calculation off?

Also, if the whole cluster is stuck, I think operators will be okay with consuming some bandwidth to quickly bring the cluster back up. If sending 81k slot information means they actually don't need to repair all the slots, maybe we would even save more on the repair bandwidth.

I think the "repair a batch at a time backwards" method you mentioned sounds interesting, but you never know what random validators can throw at you, so a few validators throwing you slots into the far future can make you really busy at repair for quite some time.

Reply:
@behzad assume the worst, turbine / repair is congested (similar to last outage) this protocol will always find the OC restart slot, which is prone to human error

Reply:
@AshwinSekar gossip has a lot more load than turbine and repair combined. I don't think we can solve congestion on one protocol by adding more load to the one already more overloaded.

@wen-coding When this was initially discussed we were talking about only 2 new crds value. This commit itself is adding 5 new pretty big values where each one fills up a packet.
It also has some overlap with already existing repair orphan request and I am not convinced this duplication will buy us anything in practice. 
The design that you cannot tell slots.last() is not the parent of last_voted_slot and you rely on receiving all crds values is also not ideal because gossip does not provide such guarantees.
Can you revisit the design addressing these issues.

Reply:
Thought about it more, I think you had a point about we probably do not need to send 81k slots in most cases. Reading previous outage reports, validators around the world normally entered the outage at about the same time so the difference of their roots is normally much smaller than duration of the outages.

That said, I still want to play it safe and send enough slots to rescue the occasional outliers back, so I propose lowering the original 81k number to MAX_SLOTS_PER_ENTRY which is 16384. This lowers the number of packets per validator from 5 to 1, and the code handling the packet don't need to deal with holes in between any more.

@carllin @mvines @t-nelson Are you okay with lowering 81k slots to 16k slots given the reasoning above?

@behzadnouri are you okay with it given we are only adding 1 RestartLastVotedForkSlots per validator now?

regarding "some overlap with already existing repair orphan request", I think:
1. If we are in an outage, apparently that's because repair orphan request didn't save us
2. Repair orphan request can figure out parent relationship between blocks and repair missing blocks on the fork, but it's **not** a global view so it **can't** tell us whether some blocks can be ignored. To quickly reach consensus between validators you do need a global view, which is something the current repair service can't give you, you do need some sort of global consensus to make validators agree on something again.

I'd be happy to organize a quick meeting if we can't reach agreement over github reviews.

Reply:
I think thats ok, with only one single such crds value the parent is also inferred, which should address @behzadnouri's concern